modules/openxr/extensions/openxr_debug_utils_extension.cpp

```diff
@@ -0,0 +1,168 @@
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "openxr_debug_utils_extension.h"
+
+#include "../openxr_api.h"
+#include "core/config/project_settings.h"
+#include "core/string/print_string.h"
+
+#include <openxr/openxr.h>
+
+OpenXRDebugUtilsExtension *OpenXRDebugUtilsExtension::singleton = nullptr;
+
+OpenXRDebugUtilsExtension *OpenXRDebugUtilsExtension::get_singleton() {
+	return singleton;
+}
+
+OpenXRDebugUtilsExtension::OpenXRDebugUtilsExtension() {
+	singleton = this;
+}
+
+OpenXRDebugUtilsExtension::~OpenXRDebugUtilsExtension() {
+	singleton = nullptr;
+}
+
+HashMap<String, bool *> OpenXRDebugUtilsExtension::get_requested_extensions() {
+	HashMap<String, bool *> request_extensions;
+
+	request_extensions[XR_EXT_DEBUG_UTILS_EXTENSION_NAME] = &debug_utils_ext;
+
+	return request_extensions;
+}
+
+void OpenXRDebugUtilsExtension::on_instance_created(const XrInstance p_instance) {
+	if (debug_utils_ext) {
+		EXT_INIT_XR_FUNC(xrCreateDebugUtilsMessengerEXT);
+		EXT_INIT_XR_FUNC(xrDestroyDebugUtilsMessengerEXT);
+
+		debug_utils_ext = xrCreateDebugUtilsMessengerEXT_ptr && xrDestroyDebugUtilsMessengerEXT_ptr;
+	}
+
+	// On successful init, setup our default messenger.
+	if (debug_utils_ext) {
+		int max_severity = GLOBAL_GET("xr/openxr/extensions/debug_utils");
+		XrDebugUtilsMessageSeverityFlagsEXT message_severities = 0;
+
+		if (max_severity >= 1) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT;
+		}
+		if (max_severity >= 2) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT;
+		}
+		if (max_severity >= 3) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT;
+		}
+		if (max_severity >= 4) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT;
+		}
+
+		XrDebugUtilsMessengerCreateInfoEXT callback_info = {
+			XR_TYPE_DEBUG_UTILS_MESSENGER_CREATE_INFO_EXT, // type
+			nullptr, // next
+			message_severities, // messageSeverities
+			XR_DEBUG_UTILS_MESSAGE_TYPE_GENERAL_BIT_EXT | XR_DEBUG_UTILS_MESSAGE_TYPE_VALIDATION_BIT_EXT | XR_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT | XR_DEBUG_UTILS_MESSAGE_TYPE_CONFORMANCE_BIT_EXT, // messageTypes
+			&OpenXRDebugUtilsExtension::_debug_callback, // userCallback
+			nullptr, // userData
+		};
+
+		xrCreateDebugUtilsMessengerEXT(p_instance, &callback_info, &default_messenger);
+	}
+}
+
+void OpenXRDebugUtilsExtension::on_instance_destroyed() {
+	if (default_messenger != XR_NULL_HANDLE) {
+		xrDestroyDebugUtilsMessengerEXT(default_messenger);
+		default_messenger = XR_NULL_HANDLE;
+	}
+
+	xrCreateDebugUtilsMessengerEXT_ptr = nullptr;
+	xrDestroyDebugUtilsMessengerEXT_ptr = nullptr;
```

Comment:
You should also reset `debug_utils_ext`.

Reply:
Do we rerun checking the extensions? I thought we only do that once.

Reply:
I've added it regardless, seeing we don't recreate instances I think it'll be safe.